modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicOptionsParameterDeclarationEnricher.java

```diff
@@ -0,0 +1,231 @@
+import static java.util.Optional.empty;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getAnnotatedElement;
+import org.mule.metadata.api.ClassTypeLoader;
+import org.mule.runtime.api.meta.model.ModelProperty;
+import org.mule.runtime.api.meta.model.declaration.fluent.BaseDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ConfigurationDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ConnectionProviderDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.OperationDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterGroupDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.ParameterizedDeclaration;
+import org.mule.runtime.api.meta.model.declaration.fluent.SourceDeclaration;
+import org.mule.runtime.extension.api.annotation.options.DynamicOptions;
+import org.mule.runtime.extension.api.annotation.options.OptionsPart;
+import org.mule.runtime.extension.api.annotation.param.Config;
+import org.mule.runtime.extension.api.annotation.param.Connection;
+import org.mule.runtime.extension.api.annotation.param.Parameter;
+import org.mule.runtime.extension.api.declaration.fluent.util.IdempotentDeclarationWalker;
+import org.mule.runtime.extension.api.declaration.type.ExtensionsTypeLoaderFactory;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.extension.api.options.OptionsResolver;
+import org.mule.runtime.extension.internal.property.OptionPartModelProperty;
+import org.mule.runtime.extension.internal.property.OptionsModelProperty;
+import org.mule.runtime.extension.internal.property.OptionsModelProperty.OptionsModelPropertyBuilder;
+import org.mule.runtime.module.extension.internal.loader.java.property.ImplementingTypeModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.OptionsResolverFactoryModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.OptionsResolverFactoryModelProperty.OptionsResolverFactoryModelPropertyBuilder;
+import org.mule.runtime.module.extension.internal.loader.java.property.ParameterGroupModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.type.ExtensionParameter;
+import org.mule.runtime.module.extension.internal.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.internal.loader.java.type.runtime.ParameterizableTypeWrapper;
+
+import java.lang.annotation.Annotation;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Some enricher
+ *
+ * @since 4.0
+ */
+public class DynamicOptionsParameterDeclarationEnricher extends AbstractAnnotatedDeclarationEnricher {
```

Comment:
need to completely review this class. Some of these things go beyond enriching and are potential signs of things being in the wrong place